third_party/nvfuser/csrc/scheduler/reduction_utils.cpp

```diff
@@ -511,6 +521,143 @@ void multiReductionInliner(
   inlineMost();
 }
 
+void propagateTransformation(
+    TensorView* reference_tv,
+    const std::unordered_set<TensorView*>& boundaryNodesSet) {
+  // set an internal boundary to allow different parallel patterns in the fusion
+  struct InternalBoundarySelector : public MaxInfoSpanningTree::Selector {
```

Comment:
This seems pretty similar to `SetSelector`, which only propagates to a specified node set, whereas here we want to stop propagation at a specified node set. Let's move this to `maxinfo_propagator.h` as that seems more consistent. 

BTW, what do you mean by an internal boundary?

Reply:
moved. I borrowed it from CFD, where fluid flow is usually blocked. Here the propagation is blocked at `boundaryNodesSet`